### PR TITLE
Testclusters: implement starting, waiting for and stopping single cluster nodes 

### PR DESCRIPTION
**Pull request by alpar-t**
The bulk of the request is new functionality in `ElasticsearchNode`. 
Note that this offers 0 configuration, this ability will be added as we start to use testclsuters and  find a need for things to be configurable to keep these to a minimum. 
My thinking is that we can auto configure in some of the cases where we have configuration for cluster-formation  ( e.x. detect that security is installed and do the right checks automatically).

The accompanying tests are still the only users of this functionality.

A quick map to help navigate the PR:
  -  setting up a directory layout and writing config file `writeConfigurationFile`
        - we don't copy the entire distribution, only the config folder and create folders in a different location to store data, logs, etc. 
  -  We start the process with `ProcessBuilder` and redirect the streams to files `start`  
     -  There are cleanup hooks to make sure processes started by the build won't  outlive it  `registerCleanupHooks`
  -  Stopping the cluster stops all processes with the same api in `stop` and reads the files into the stream on failure.
  -  We wait for the cluster to be ready only when and each time we get a request for it's address.  There's a generic mechanism to deal with waiting, but it's not configurable for now. 

Tests are passing on Windows.

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 1]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
+            );
+        }
+        try {
+            processBuilder.directory(workingDir);
+            Map<String, String> environment = processBuilder.environment();
+            // Don't inherit anything from the environment for as that would  lack reproductability
+            environment.clear();
+            if (javaHome != null) {
+                environment.put("JAVA_HOME", getJavaHome().getAbsolutePath());
+            } else if (System.getenv().get("JAVA_HOME") != null) {
```
Comment (original-brownbear) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Wouldn't it be better to get this from the project like we do currently (`project.runtimeJavaHome`)?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 2]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
+            // don't buffer all in memory, make sure we don't block on the default pipes
+            processBuilder.redirectError(ProcessBuilder.Redirect.appendTo(getStdErrFile()));
+            processBuilder.redirectOutput(ProcessBuilder.Redirect.appendTo(getStdoutFile()));
+            esProcess = processBuilder.start();
+        } catch (IOException e) {
+            throw new TestClustersException("Failed to start ES process for " + this, e);
+        }
+    }
+
+    private void registerCleanupHooks() {
```
Comment (original-brownbear) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Why are we doing our own thing here to ensure this runs instead of simply using the `afterTask` hook?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 3]
```diff
@@ -83,21 +386,122 @@ private void checkFrozen() {
...
+                        lastException = e;
+                    } else {
+                        e.addSuppressed(lastException);
+                        lastException = e;
+                    }
+                }
+                try {
+                    Thread.sleep(500);
+                }
+                catch (TestClustersException e) {
```
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: We want to keep trying the predicate if an exception occurs, except for `TestClustersException` which we want to bubble up directly. 
It's a way of saying something went wrong while waiting, there's no way it will recover and we want to fail fast. I'll fix the formatting thanks!

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 2 above)
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: We do both. We do stop clusters in the `afterTask` hook.
This is for when we don't get that far. 
e.x  Ctrl^C on the console, etc

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 4]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
         logger.info("Stopping `{}`, tailLogs: {}", this, tailLogs);
+        if (esProcess == null && tailLogs) {
+            // This is a special case. If start() throws an exception the plugin will still call stop
+            // Another exception here would eat the orriginal.
+            return;
+        }
+        requireNonNull(esProcess, "Can't stop `" + this + "` as it was not started or already stopped.");
+        stopHandle(esProcess.toHandle());
+        if (tailLogs) {
+            logFileContents("Standard output of node", getStdoutFile());
```
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: I think that's something to keep in mind for when we start using this. 
This problem would also be solved by build scans and it's different enough from what we have that I would like to run into it first and then think about solutions.

**Comment by atorok** [comment]
Thanks for the review @original-brownbear !  I left some answers.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: sure sounds good :)

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Ah wait :) doesn't that `try`- `catch` only cover `Thread.sleep` here anyway which will never throw `TestClustersException`?

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 2 above)
Comment (original-brownbear) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Ah, makes sense :)

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 5]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
+            processBuilder.redirectOutput(ProcessBuilder.Redirect.appendTo(getStdoutFile()));
+            esProcess = processBuilder.start();
+        } catch (IOException e) {
+            throw new TestClustersException("Failed to start ES process for " + this, e);
+        }
+    }
+
+    private void registerCleanupHooks() {
+        requireNonNull(esProcess, "Attempted to register cleanup hooks without a process");
+        // The Gradle daemon interrupts all threads from the build when a build concludes,
```
Comment (original-brownbear) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: I just tried this locally, the daemon doesn't seem to interrupt this kind of `Thread` if the build concludes successfully. So we'd be leaking a `Thread` here for each ES node we spin up when using the daemon?
Maybe a better Pattern would be to just use a single `Thread` for this globally, keep a static reference to it and a collection that holds all the running Es nodes and then have that thread stop all nodes in that collection on interrupt?

**Comment by original-brownbear** [comment]
@atorok thanks! Looks good in general :), but see my comment about the thread leak.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 6]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
+                environment.put("JAVA_HOME", System.getenv().get("JAVA_HOME"));
+            } else {
+                logger.warn("{}: No javaHome configured, will rely on default java detection", this);
+            }
+            environment.put("ES_PATH_CONF", getConfigFile().getParentFile().getAbsolutePath());
+            environment.put("ES_JAVA_OPTIONS", "-Xms512m -Xmx512m");
+            // don't buffer all in memory, make sure we don't block on the default pipes
+            processBuilder.redirectError(ProcessBuilder.Redirect.appendTo(getStdErrFile()));
+            processBuilder.redirectOutput(ProcessBuilder.Redirect.appendTo(getStdoutFile()));
+            esProcess = processBuilder.start();
```
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: It's not because we do our own cleanup.
It's hard to write a test that would assert that no ES processes are left after executing,
maybe we could do it in an after build hook that always runs if we really wanted to, but  I always check for them when developing and I haven't found any with this implementation so far. 
I have ran into cases where the ant runner left processes behind too.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 7]
```diff
@@ -59,22 +102,282 @@ public Distribution getDistribution() {
...
+    }
+
+    private File getStdErrFile() {
+        return new File(getConfPathLogs(), "es.stderr.log");
+    }
+
+    private void writeConfigFile() {
+        getConfigFile().getParentFile().mkdirs();
+        getConfPathRepo().mkdirs();
+        getConfPathData().mkdirs();
```
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: you're right, I changed the name.

**Comment by atorok** [comment]
I moved the cleanup of processes outside of `ElasticsearchNode` I think it's a better fit for the plugin, and this also takes care to make sure the thread is not leaked in the daemon.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java [hunk 8]
```diff
@@ -272,6 +283,54 @@ private void autoConfigureClusterDependencies(
...
+        synchronized (runningClusters) {
+            if (executorService == null || executorService.isTerminated()) {
+                executorService = Executors.newSingleThreadExecutor();
+            } else {
+                throw new IllegalStateException("Trying to configure executor service twice");
+            }
+        }
+        // When the Gradle daemon is used, it will interrupt all threads when the build concludes.
+        executorService.submit(() -> {
+            while (true) {
```
Comment (original-brownbear) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: NIT: the `while` loop is redundant when we sleep forever anyway I think :)

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java [hunk 9]
```diff
@@ -272,6 +283,54 @@ private void autoConfigureClusterDependencies(
...
+        // When the Daemon is not used, or runs into issues, rely on a shutdown hook
+        // When the daemon is used, but does not work correctly and eventually dies off (e.x. due to non interruptable
+        // thread in the build) process will be stopped eventually when the daemon dies.
+        Runtime.getRuntime().addShutdownHook(new Thread(TestClustersPlugin::shutDownAllClusters));
+    }
+
+    private static void shutdownExecutorService() {
+        executorService.shutdownNow();
+        try {
+            executorService.awaitTermination(1, TimeUnit.MINUTES);
```
Comment (original-brownbear) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: Maybe log an error (or even fail more visibly) if this returns `false` and didn't actually finish terminating all the nodes?

Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: good catch, thanks !

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java (hunk 8 above)
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: I was thinking about that, but I think it conveys the intent better.

Comment (original-brownbear) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java]: sure, makes sense :)

**Comment by atorok** [comment]
@nik9000 can you take another look please.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java [hunk 10]
```diff
@@ -20,25 +20,67 @@
...
 
-    public ElasticsearchNode(String name, GradleServicesAdapter services) {
+    ElasticsearchNode(String path, String name, GradleServicesAdapter services, File artifactsExtractDir, File workingDirBase) {
+        this.path = path;
         this.name = name;
         this.services = services;
+        this.artifactsExtractDir = artifactsExtractDir;
+        this.workingDir = new File(workingDirBase, safeName(name));
+        this.waitConditions = new LinkedHashMap<>();
+        waitConditions.put("http ports file",  (node -> node.getHttpPortsFile().exists()));
```
Comment (nik9000) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: I don't think we need the `(` wrapping the function.

File: buildSrc/build.gradle [hunk 11]
```diff
@@ -163,6 +163,10 @@ if (project != rootProject) {
   apply plugin: 'nebula.maven-base-publish'
   apply plugin: 'nebula.maven-scm'
 
+  // we need to apply these again to override the build plugin
+  targetCompatibility = "10"
```
Comment (nik9000) [inline buildSrc/build.gradle]: Is this a separate change from this? Like, is it just required because of this change?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy [hunk 12]
```diff
@@ -212,7 +212,7 @@ class PrecommitTasks {
     private static Task configureNamingConventions(Project project) {
         if (project.sourceSets.findByName("test")) {
             Task namingConventionsTask = project.tasks.create('namingConventions', NamingConventionsTask)
-            namingConventionsTask.javaHome = project.runtimeJavaHome
+            namingConventionsTask.javaHome = project.compilerJavaHome
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: Is this related to the testclusters change?

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/NamingConventionsCheck.java [hunk 13]
```diff
@@ -17,7 +17,7 @@
  * under the License.
  */
 
-package org.elasticsearch.test;
+package org.elasticsearch.gradle.test;
```
Comment (nik9000) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/NamingConventionsCheck.java]: Maybe this package move should be in its own PR.

File: test/framework/build.gradle [hunk 14]
```diff
@@ -64,7 +64,7 @@ thirdPartyAudit.excludes = [
 
 task namingConventionsMain(type: org.elasticsearch.gradle.precommit.NamingConventionsTask) {
   checkForTestsInMain = true
-  javaHome = project.runtimeJavaHome
+  javaHome = project.compilerJavaHome
```
Comment (nik9000) [inline test/framework/build.gradle]: Is this related to the change?

**Review by nik9000 — COMMENTED** [review]
I'm happy with this. I think it pushes the project forwards and i like it. I *think* there are a couple of unrelated changes bundled up in it. At least, I can't see how they are related. Could you explain them? Or maybe move them into their own PR so they get a nice commit message explaining them?

**Comment by atorok** [comment]
@nik9000 I should have offered more details on these changes . I now realize they seem out of place. They do need to be part of this PR.
In #34746 I moved buildSrc to target java 10 ( 11 doesn't work with gradle 4.10 ) but this is the first change that actually uses something newer than java 8. Turns out my previous change was incomplete as it changed `buildSrc` but not `build-tools`  because the build plugin was setting it back to 8 so we need to be specific about it again after applying the plugin. I'll look to clean up the package naming change, that's unintentional.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Indeed, it looks like this only covers the sleep.

**Comment by atorok** [comment]
@nik9000 we do also need to run the naming convention check with the compiler java as build-tools how has classes for newer java versions.  We could do it locally for build tools only but it doesn't matter much since we'll be using compiler java for this check with  `testingConventions` soon anyhow.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: Fixed it, thanks !

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 5 above)
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: This is now fixed, thanks!

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: We don't have a way to easily test this right now. Will consider adding something lather on if this turns out to be problematic.

File: buildSrc/build.gradle (hunk 11 above)
Comment (atorok) [reply buildSrc/build.gradle]: explained below

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/NamingConventionsCheck.java (hunk 13 above)
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/NamingConventionsCheck.java]: reverted

File: test/framework/build.gradle (hunk 14 above)
Comment (atorok) [reply test/framework/build.gradle]: yes, explained bellow

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy (hunk 12 above)
Comment (atorok) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: yes, explained bellow

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java (hunk 10 above)
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java]: fixed, thanks

**Comment by atorok** [comment]
@nik9000 I double checked to be sure. All comments on the PR should be addressed.

**Comment by atorok** [comment]
@elasticmachine run default distro tests

**Comment by atorok** [comment]
@elasticmachine run default distro tests

**Review by nik9000 — APPROVED** [review]
